src/backend/distributed/executor/distributed_intermediate_results.c

```diff
@@ -0,0 +1,320 @@
+ *
+ * Result is list of DistributedResultFragment, each of which represents a
+ * partition of results. Empty results are omitted. Therefore, if we have N tasks
+ * and target relation has M shards, we will have NxM-(number of empty results)
+ * fragments.
+ */
+List *
+PartitionTasklistResults(char *resultIdPrefix, List *selectTaskList,
+						 DistTableCacheEntry *targetRelation,
+						 bool binaryFormat)
+{
+	/*
+	 * Make sure that this transaction has a distributed transaction ID.
+	 *
+	 * Intermediate results will be stored in a directory that is derived
+	 * from the distributed transaction ID.
+	 */
+	UseCoordinatedTransaction();
+
+	WrapTasksForPartitioning(resultIdPrefix, selectTaskList, targetRelation,
+							 binaryFormat);
+	return ExecutePartitionTaskList(selectTaskList, targetRelation);
+}
+
+
+/*
+ * WrapTasksForPartitioning wraps the query for each of the tasks by a call
+ * to worker_partition_query_result(). Target list of the wrapped query should
+ * match the tuple descriptor in ExecutePartitionTaskList().
+ */
+static void
+WrapTasksForPartitioning(char *resultIdPrefix, List *selectTaskList,
+						 DistTableCacheEntry *targetRelation,
+						 bool binaryFormat)
+{
+	ListCell *taskCell = NULL;
+	ShardInterval **shardIntervalArray = targetRelation->sortedShardIntervalArray;
+	int shardCount = targetRelation->shardIntervalArrayLength;
+
+	ArrayType *minValueArray = NULL;
+	ArrayType *maxValueArray = NULL;
+	Var *partitionColumn = targetRelation->partitionColumn;
+	int partitionColumnIndex = partitionColumn->varoattno - 1;
+	Oid intervalTypeId = partitionColumn->vartype;
+	int32 intervalTypeMod = partitionColumn->vartypmod;
+	Oid intervalTypeOutFunc = InvalidOid;
+	bool intervalTypeVarlena = false;
+	getTypeOutputInfo(intervalTypeId, &intervalTypeOutFunc, &intervalTypeVarlena);
+
+	ShardMinMaxValueArrays(shardIntervalArray, shardCount, intervalTypeOutFunc,
+						   &minValueArray, &maxValueArray);
+	StringInfo minValuesString = ArrayObjectToString(minValueArray, TEXTOID,
+													 intervalTypeMod);
+	StringInfo maxValuesString = ArrayObjectToString(maxValueArray, TEXTOID,
+													 intervalTypeMod);
+
+	foreach(taskCell, selectTaskList)
+	{
+		Task *selectTask = (Task *) lfirst(taskCell);
+		StringInfo wrappedQuery = makeStringInfo();
+		List *shardPlacementList = selectTask->taskPlacementList;
+
+		ShardPlacement *shardPlacement = linitial(shardPlacementList);
+		char *taskPrefix = SourceShardPrefix(resultIdPrefix, selectTask->anchorShardId);
+		char *partitionMethodString = targetRelation->partitionMethod == 'h' ?
+									  "hash" : "range";
+		const char *binaryFormatString = binaryFormat ? "true" : "false";
+
+		appendStringInfo(wrappedQuery,
+						 "SELECT %d, partition_index"
+						 ", %s || '_' || partition_index::text "
+						 ", rows_written "
+						 "FROM worker_partition_query_result"
+						 "(%s,%s,%d,%s,%s,%s,%s) WHERE rows_written > 0",
+						 shardPlacement->nodeId,
```

Comment:
I wonder whether there's another way in which can fill in the node ID, such that we can fail over to another replica if needed and return a different node ID in the fragment list.

One way could be to look at the task execution state in the DistributedExecution. We'd probably have to include the anchor shard ID in the result and map that back to a ShardCommandExecution to figure out which node the task was executed on, which probably requires a bunch of refactoring since the ShardCommandExecutions are not easily accessible from the DistributedExecution data structure. 

Anyway, we should probably error out here when there are multiple placements if we don't properly support it.

Reply:
What is your take on something like https://github.com/citusdata/citus/pull/3363? I haven't tested it much yet, just opened it to get early feedback.